### ISSUE TITLE
[FR] Native flavor for diffusion model adapters (LoRA, LoKr, LoHa)

### ISSUE THREAD

**Issue by Rasaboun** [mlflow/mlflow#22122]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

MLflow has no way to log and register diffusion model adapters (LoRA, LoKr, LoHa) as first-class models. These are the standard way to fine-tune Stable Diffusion, SDXL, and Flux models, but they don't fit any existing MLflow flavor because they're not standalone models — they're lightweight weight adapters that get applied on top of a base model at inference time.

A native `diffusers-adapter` flavor (or extending the existing `transformers` flavor) would let users log, version, and deploy LoRA models the same way they do with any other ML model in MLflow.

### Motivation

> #### What is the use case for this feature?

Fine-tuning diffusion models with LoRA is one of the most common workflows in generative AI right now. Tools like ai-toolkit, kohya-ss, and diffusers all produce LoRA weights as `.safetensors` files. Users need to track which LoRA version came from which training run, compare them, and eventually serve them.

> #### Why is this use case valuable to support for MLflow users in general?

LoRA fine-tuning is how most teams customize image generation models. The ecosystem is huge — Civitai alone hosts hundreds of thousands of LoRA models. Anyone using MLflow for diffusion model training hits the same wall: you can track metrics and log images, but the actual model output (the LoRA weights) is just a plain artifact with no model registry support.

> #### Why is this use case valuable to support for your project(s) or organization?

We're building an MLflow integration for ostris/ai-toolkit (LoRA training for Stable Diffusion, SDXL, Flux). Everything works except model registration — the LoRA weights don't show up in the Models section because there's no flavor that understands adapter weights.

> #### Why is it currently difficult to achieve this use case?

You can't even register a LoRA as a model without a workaround. `mlflow.register_model()` on a raw artifact fails on MLflow 3.x with `Unable to find a logged_model` — it requires a `log_model()` call, not just `log_artifact()`. So you're forced to create a pyfunc stub just to get into the Model Registry:

```python
class LoRAStub(mlflow.pyfunc.PythonModel):
    def predict(self, context, model_input):
        raise NotImplementedError("Use diffusers to load this LoRA adapter")

mlflow.pyfunc.log_model(
    name="lora_model",
    python_model=LoRAStub(),
    artifacts={"lora_weights": "output/lora.safetensors"},
    registered_model_name="my-lora",
)
```

This works for registration and lineage but the model can't actually be served. Beyond that, even if you tried to write a real `predict()`:

- You have to write a custom model class that handles loading the base model + applying the adapter
- The base model (often 5-10GB) can't reasonably be bundled as an artifact, so you end up storing a reference string and hoping it resolves at load time
- Different adapter formats (safetensors, kohya-ss, diffusers) need different loading logic
- The `predict()` signature doesn't map well to image generation (prompts in, images out, with dozens of generation params like guidance_scale, num_inference_steps, etc.)
- No standardized way to declare which base model the adapter was trained on

A native flavor would handle all of this with a clean API like:
```python
mlflow.diffusers.log_adapter(
    adapter_path="output/lora.safetensors",
    base_model="black-forest-labs/FLUX.1-dev",
    adapter_type="lora",
    artifact_path="lora_model",
)
```

### Details

A native flavor would need to handle:

- **Adapter types**: LoRA (most common), LoKr, LoHa, full fine-tune diffs
- **Formats**: safetensors (standard), kohya-ss state dict format, diffusers format
- **Base model reference**: store which base model the adapter was trained on without bundling the full model weights
- **Loading**: use `diffusers` library to load base model + apply adapter via `pipe.load_lora_weights()`
- **Inference signature**: prompts + generation params → images
- **Model card metadata**: adapter rank, training dataset, base model, etc.

This could either be a new `mlflow.diffusers_adapter` flavor or an extension of the existing `mlflow.transformers` flavor since diffusers is part of the Hugging Face ecosystem.

### What machine learning domain(s) is this feature request about?

- `domain/deep-learning`

### What area(s) of MLflow is this feature request about?

- `area/models`

**Comment by TomeHirata** [comment]
We already support PEFT [as part of the transformers flavor](https://mlflow.org/docs/latest/ml/deep-learning/transformers/tutorials/fine-tuning/transformers-peft/), is this sufficient?

**Comment by Rasaboun** [comment]
I looked at the transformers PEFT support and it works well for text models but doesn't cover diffusion models

The transformers flavor expects transformers.Pipeline and PreTrainedModel so diffusers pipelines like StableDiffusionPipeline or FluxPipeline can't go through it, diffusers also has its own way of loading LoRA adapters via load_lora_weights() which isn't handled

Happy to contribute if this is worth supporting

**Comment by TomeHirata** [comment]
Please feel free to extend the existing PEFT support for diffusion models, but can we update the issue title/description?

**Comment by Rasaboun** [comment]
Should I extend mlflow.transformers to also handle diffusers pipelines or create a separate mlflow.diffusers flavor ?

I think a separate flavor would be cleaner since diffusers has its own pipeline types, LoRA loading mechanism and inference flow (prompt → image) so fitting it into the existing transformers code would mean adding parallel paths everywhere

**Comment by TomeHirata** [comment]
Let me look into the code structure of diffusers, and get back to you

**Comment by Rasaboun** [comment]
I looked into extending the transformers flavor but DiffusionPipeline and transformers.Pipeline are separate class hierarchies with no shared interface

DiffusionPipeline is a container for multiple models (unet, vae, text_encoder, scheduler) rather than a wrapper around a single model. It has no `.task` property and no `.model` attribute. The save format is also different with each component getting its own subfolder. Adapter loading works through `pipe.load_lora_weights(path)` which injects weights into the unet and text_encoder in place rather than wrapping through `PeftModel.from_pretrained()` like the existing PEFT support does

Fitting this into the transformers flavor would mean parallel code paths through save_model, load_model, flavor_config and the predict wrapper that share nothing with the existing logic

I went with a standalone `mlflow.diffusers` flavor instead. It follows the same patterns as sklearn/transformers and stores adapter weights as the artifact with the base model as a reference string since base models are 5-10GB. The pyfunc wrapper takes a prompt and generation params and returns PNG bytes. Tests pass on CPU (save/load round-trip, log_model, pyfunc wrapper, signature, model registry, error handling)

I will push the branch if you want to take a look
Open to doing it differently if you have something else in mind

**Comment by Rasaboun** [comment]
Hey @TomeHirata, I put together an implementation : #22219 (PR is closed but the code and commits are readable)

Standalone `mlflow.diffusers` flavor following the sklearn/transformers patterns with `save_model`, `log_model`, `load_model` and pyfunc `predict()`, 21 tests passing and a demo script included

Tested end-to-end with a real FLUX.2 Klein 4B LoRA fine-tune on GPU including checkpoint logging and artifact verification through the MLflow API

Let me know if the approach works or if you'd rather go a different direction, I can reopen as a proper PR once we're aligned on the design

**Comment by TomeHirata** [comment]
Thanks for putting together the implementation. In this case, a dedicated diffuser flavor makes sense, I've just attached "ready" label. Please feel free to reopen the PR.
